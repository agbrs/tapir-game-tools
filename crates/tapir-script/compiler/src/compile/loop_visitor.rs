use crate::{
    ast::{Function, Statement, StatementKind},
    reporting::{CompilerErrorKind, Diagnostics},
};

#[derive(Debug)]
pub struct LoopContainsNoBreak;

pub fn visit_loop_check(function: &mut Function<'_>, diagnostics: &mut Diagnostics) {
    visit_block_in_loop(&mut function.statements, false, diagnostics);
}

#[derive(Debug, PartialEq, Eq)]
enum LoopReturn {
    ContainsNoBreaks,
    MayBreak,
}

fn visit_block_in_loop(
    block: &mut [Statement<'_>],
    is_in_loop: bool,
    diagnostics: &mut Diagnostics,
) -> LoopReturn {
    let mut result = LoopReturn::ContainsNoBreaks;

    for statement in block {
        match &mut statement.kind {
            StatementKind::Error
            | StatementKind::VariableDeclaration { .. }
            | StatementKind::Assignment { .. }
            | StatementKind::Wait
            | StatementKind::Nop
            | StatementKind::Call { .. }
            | StatementKind::Spawn { .. } => {}
            StatementKind::Return { .. } => {
                // Not returning to ensure that we catch continues and break outside of loops
            }
            StatementKind::Continue => {
                if !is_in_loop {
                    diagnostics.add_message(
                        CompilerErrorKind::BreakOrContinueOutsideOfLoop
                            .into_message(statement.span),
                    );
                }
            }
            StatementKind::Break => {
                if !is_in_loop {
                    diagnostics.add_message(
                        CompilerErrorKind::BreakOrContinueOutsideOfLoop
                            .into_message(statement.span),
                    );
                }

                result = LoopReturn::MayBreak;
            }
            StatementKind::If {
                true_block,
                false_block,
                ..
            } => {
                let true_case = visit_block_in_loop(true_block, is_in_loop, diagnostics);
                let false_case = visit_block_in_loop(false_block, is_in_loop, diagnostics);
                if true_case == LoopReturn::MayBreak || false_case == LoopReturn::MayBreak {
                    result = LoopReturn::MayBreak;
                }
            }
            StatementKind::Loop { block } => {
                if visit_block_in_loop(block, true, diagnostics) == LoopReturn::ContainsNoBreaks {
                    statement.meta.set(LoopContainsNoBreak);
                }
            }
        }
    }

    result
}

#[cfg(test)]
mod test {
    use std::fs;

    use insta::{assert_ron_snapshot, assert_snapshot, glob};

    use crate::{grammar, lexer::Lexer, tokens::FileId};

    use super::*;

    #[test]
    fn symtab_success_snapshot_tests() {
        glob!("snapshot_tests", "loop_visitor/*_success.tapir", |path| {
            let input = fs::read_to_string(path).unwrap();

            let lexer = Lexer::new(&input, FileId::new(0));
            let parser = grammar::ScriptParser::new();
            let file_id = FileId::new(0);

            let mut diagnostics = Diagnostics::new(file_id, path.file_name().unwrap(), &input);

            let mut script = parser.parse(file_id, &mut diagnostics, lexer).unwrap();

            for function in &mut script.functions {
                visit_loop_check(function, &mut diagnostics);
            }

            assert_ron_snapshot!(script, {
                ".**.span" => "[span]",
            });
        });
    }

    #[test]
    fn symtab_fail_snapshot_tests() {
        glob!("snapshot_tests", "loop_visitor/*_fail.tapir", |path| {
            let input = fs::read_to_string(path).unwrap();

            let lexer = Lexer::new(&input, FileId::new(0));
            let parser = grammar::ScriptParser::new();
            let file_id = FileId::new(0);

            let mut diagnostics = Diagnostics::new(file_id, path.file_name().unwrap(), &input);

            let mut script = parser.parse(file_id, &mut diagnostics, lexer).unwrap();

            for function in &mut script.functions {
                visit_loop_check(function, &mut diagnostics);
            }

            assert_snapshot!(diagnostics.pretty_string(false));
        });
    }
}
