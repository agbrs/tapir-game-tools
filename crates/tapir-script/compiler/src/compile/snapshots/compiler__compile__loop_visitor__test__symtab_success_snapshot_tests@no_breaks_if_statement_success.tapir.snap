---
source: crates/tapir-script/compiler/src/compile/loop_visitor.rs
expression: script
input_file: crates/tapir-script/compiler/src/compile/snapshot_tests/loop_visitor/no_breaks_if_statement_success.tapir
---
Script(
  functions: [
    Function(
      name: "@toplevel",
      span: "[span]",
      statements: [
        Statement(
          span: "[span]",
          kind: Loop(
            block: [
              Statement(
                span: "[span]",
                kind: If(
                  condition: Expression(
                    span: "[span]",
                    kind: Bool(true),
                    meta: {},
                  ),
                  true_block: [
                    Statement(
                      span: "[span]",
                      kind: Continue,
                      meta: {},
                    ),
                  ],
                  false_block: [
                    Statement(
                      span: "[span]",
                      kind: Return(
                        values: [],
                      ),
                      meta: {},
                    ),
                  ],
                ),
                meta: {},
              ),
            ],
          ),
          meta: {
            "compiler::compile::loop_visitor::LoopContainsNoBreak": "LoopContainsNoBreak",
          },
        ),
      ],
      arguments: [],
      return_types: FunctionReturn(
        types: [],
        span: "[span]",
      ),
      modifiers: FunctionModifiers(
        is_event_handler: None,
      ),
    ),
  ],
)
